### PR TITLE
Fix broken LegacyDFSCluster::create()

### PR DESCRIPTION
**Pull request by glye**
> Fixes https://jira.ez.no/browse/EZP-26760
> Moved here from https://github.com/ezsystems/ezpublish-kernel/pull/1825
> Ready for review

Fixes some issues found with `LegacyDFSCluster::create()` during work on the cluster migrator.
- Doc: It doesn't throw InvalidArgumentException on existing entry, it updates it instead.
- Doc: It doesn't throw BadStateException on zero rowcount anymore, as that is unreliable across DBMSs.
- Removed 'expired' from update statement, as that value isn't provided in the insert statement.
- Removed rowcount check (see above).
- Use DateTime instead of timestamp, as the BinaryFileCreateStruct->mtime docblock says.
- Fixed wrong `param` docblock.

### CODE REVIEW

File: eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php [hunk 1]
```diff
@@ -60,6 +56,11 @@ public function create(SPIBinaryFileCreateStruct $binaryFileCreateStruct)
     {
         $path = $this->addPrefix($binaryFileCreateStruct->id);
 
+        // Some code has been using integer, though DateTime is correct. Convert to DateTime if needed.
```
Comment (andrerom) [inline eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php]: shouldn't this rather be enforced if api doc says date time and so we can rather fix the use?

Comment (glye) [inline eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php]: @andrerom Ok, sure. In that case maybe we should throw an `InvalidArgumentException` if the mtime isn't a `DateTime`? Bit of a BC break. Anyway let's see if all tests pass first.

**Comment by glye** [comment]
Update: Do not convert timestamps to DateTimes, expect api doc to be followed.
Consider throwing `InvalidArgumentException` if it isn't.

File: eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php (hunk 1 above)
Comment (glye) [inline eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php]: On second thought, that is probably what you meant by "enforce" :)

**Comment by glye** [comment]
Update:
- Throw `InvalidArgumentException` if the mtime isn't a DateTime. Add test case for that.
- Fixed wrong `param` docblock.

File: eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php [hunk 2]
```diff
@@ -47,17 +47,21 @@ public function __construct(Connection $connection, UrlDecorator $urlDecorator =
...
      *
-     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException if an error occurs creating the record
+     * @throws InvalidArgumentException if the $binaryFileCreateStruct is invalid
      * @throws RuntimeException if a DBAL error occurs
      *
      * @return \eZ\Publish\SPI\IO\BinaryFile
      */
     public function create(SPIBinaryFileCreateStruct $binaryFileCreateStruct)
     {
+        if (!($binaryFileCreateStruct->mtime instanceof DateTime)) {
```
Comment (andrerom) [inline eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php]: Possible followup improvment, but you can pherhaps do it here already:
- Update doc and code here to allow [DateTimeInterface](http://php.net/manual/en/class.datetimeinterface.php) which was added in PHP 5.5 and up

**Comment by glye** [comment]
Review ping @bdunogier (requirement for the cluster migrator)

File: eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php [hunk 3]
```diff
@@ -47,17 +47,21 @@ public function __construct(Connection $connection, UrlDecorator $urlDecorator =
     /**
      * Inserts a new reference to file $spiBinaryFileId.
      *
-     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException If a file $spiBinaryFileId already exists
+     * @since 6.8 The mtime of the $binaryFileCreateStruct must be a DateTime, as specified in the struct doc.
```
Comment (andrerom) [inline eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php]: 6.10

**Comment by andrerom** [comment]
@bdunogier Could you have a look on this sometime this week to unblock the cluster script?

File: eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php (hunk 3 above)
Comment (glye) [inline eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php]: fixed

**Comment by glye** [comment]
Squashed and rebased on master.

**Comment by glye** [comment]
Merged in 6.7+: https://github.com/ezsystems/ezpublish-kernel/commit/2b74e9df15620b98f45cd31ce25efc14f1de4569

**Pull request closed without merging**